edb/schema/name.py

```diff
@@ -20,102 +20,244 @@
 from __future__ import annotations
 from typing import *
 
+import abc
 import functools
 import re
 
 from edb import errors
 
 
-NameT = TypeVar("NameT", bound="QualifiedName")
-
-
-class QualifiedName(str):
-    __slots__ = ('module', 'name')
-
-    module: str
-    name: str
-
-    def __new__(
-        cls: Type[NameT],
-        name: Union[QualifiedName, str],
-        module: Optional[str] = None,
-    ) -> NameT:
-        if not name:
-            raise NameError('name must not be empty')
-
-        if isinstance(name, QualifiedName):
-            _name = name.name
-            _module = name.module
-        elif module is not None:
-            _name = name
-            _module = module
-        else:
-            _module, _, _name = name.rpartition('::')
-
-            if not _module:
-                if not module:
-                    err = (
-                        f'improperly formed name {name!r}: '
-                        f'module is not specified'
-                    )
-                    raise errors.InvalidReferenceError(err)
-                else:
-                    _module = module
-
-        # Ignore below since Mypy doesn't believe you can pass `_fullname` to
-        # object.__new__.
-        _fullname = f"{_module}::{_name}"
-        result = super().__new__(cls, _fullname)  # type: ignore
-        result.name = _name
-        result.module = _module
-
-        return cast(NameT, result)
-
-    def as_tuple(self) -> Tuple[str, str]:
-        return (self.module, self.name)
-
-    @staticmethod
-    def is_qualified(name: Union[QualifiedName, str]) -> bool:
-        return isinstance(name, QualifiedName) or '::' in name
-
-
-class UnqualifiedName(QualifiedName):
-
-    def __new__(cls, name: str) -> UnqualifiedName:
-        # Ignore below since Mypy doesn't believe you can pass `name` to
-        # object.__new__.
-        result = str.__new__(cls, name)  # type: ignore
-        result.name = name
-        result.module = ''
-
-        return cast(UnqualifiedName, result)
-
-
-def split_name(
-    name: Union[QualifiedName, str]
-) -> Tuple[Union[QualifiedName, str], Optional[str], str]:
-    module: Optional[str]
-    nqname: str
-
-    if isinstance(name, QualifiedName):
-        module = name.module
-        nqname = name.name
-    elif isinstance(name, tuple):
-        module = name[0]
-        nqname = name[1]
-        name = module + '::' + nqname if module else nqname
-    elif QualifiedName.is_qualified(name):
-        name = QualifiedName(name)
-        module = name.module
-        nqname = name.name
-    else:
-        module = None
-        nqname = name
+NameT = TypeVar("NameT", bound="Name")
+QualifiedNameT = TypeVar("QualifiedNameT", bound="QualifiedName")
+UnqualifiedNameT = TypeVar("UnqualifiedNameT", bound="UnqualifiedName")
+
+
+if TYPE_CHECKING:
```

Comment:
Can you add a comment explaining why you have this if here? I don't get it, for example.

Reply:
Added a comment.